atc/creds/vault/vault.go

```diff
@@ -18,18 +19,18 @@ type SecretReader interface {
 // Vault converts a vault secret to our completely untyped secret
 // data.
 type Vault struct {
-	SecretReader SecretReader
-	Prefix       string
-	SharedPath   string
+	SecretReader    SecretReader
+	Prefix          string
+	LookupTemplates []*template.Template
+	SharedPath      string
 }
 
 // NewSecretLookupPaths defines how variables will be searched in the underlying secret manager
 func (v Vault) NewSecretLookupPaths(teamName string, pipelineName string, allowRootPath bool) []creds.SecretLookupPath {
 	lookupPaths := []creds.SecretLookupPath{}
-	if len(pipelineName) > 0 {
```

Comment:
This is interesting. Will the templates will just be evaluated with an empty pipeline name now? :thinking:

Reply:
Oh, good catch - I think there needs to be logic to suppress generated paths that have `//`, which is what other credential managers do in the same situation. Vault is perfectly happy to receive such paths, but they might end up pointing to a wrong place in the event I evaluate a `{{Pipeline}}`-having template in a non-`{{Pipeline}}` context.

Reply:
Logic added, same as in the other cases. This still only catches the exact `/{{.Pipeline}}/` case, as opposed to e.g. `/pipeline_{{.Pipeline}}/` or something, should anyone use that.

It might be stronger to change to a test where we probe whether the template is pipeline-name-dependent (instantiate twice with different pipeline names and see if we get different strings), and refuse to use it for real if there is no pipeline name in the current context. I'm not sure on bundling that in as a general behavior change in this PR, though.

Reply:
Maybe we could provide the template one of two structs:

```go
type PipelineEvaluation struct {
  Team     string
  Pipeline string
  Secret   string
}

type TeamEvaluation struct {
  Team   string
  Secret string
}
```

This way we can explicitly capture an error returned when the template tries to access the missing `.Pipeline` field just ignore the path. I'm not sure if there would be a nice error type to check for though.

Reply:
Were you able to try the above approach? Seems like it'd be a little better than checking for `//` since it would also catch exotic scenarios like `/team:{{.Team}}/pipeline:{{.Pipeline}}/`.

Reply:
Yes, I ended up adding something much like this at template creation time. Details in the latest commit and much rambling in its commit message.